Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,262 @@
+    /// - Returns: An optional closure to be called immediately after the new `Rendering` is produced, which takes the rendering as the only parameter.
+    func workflowWillRender<WorkflowType: Workflow>(
+        _ workflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    ) -> ((WorkflowType.Rendering) -> Void)?
+
+    /// Called after an existing `Workflow` is updated.
+    /// - Parameters:
+    ///   - oldWorkflow: The previous `Workflow`
+    ///   - newWorkflow: The new `Workflow`
+    ///   - state: The state **after** the update has occurred.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidChange<WorkflowType: Workflow>(
+        from oldWorkflow: WorkflowType,
+        to newWorkflow: WorkflowType,
+        state: WorkflowType.State,
+        session: WorkflowSession
+    )
+
+    /// Called after a `WorkflowAction` is received, but **before** it has been handled and propagated up the tree.
+    /// - Parameters:
+    ///   - action: The action that was received.
+    ///   - session: The `WorkflowSession` corresponding to the backing `WorkflowNode`.
+    func workflowDidReceiveAction<Action: WorkflowAction>(
+        _ action: Action,
+        session: WorkflowSession
+    )
+}
+
+// MARK: - WorkflowSession
+
+/// A `WorkflowSession`encapsulates the information that gives a `WorkflowNode` its identity.
+/// In particular, it captures:
+/// - The type of the corresponding `Workflow` conformance.
+/// - The `String` key used when the workflow was rendered.
+/// - An `Identifier` type that is unique across program execution.
+/// - An optional reference to a parent `WorkflowSession`, to differentiate root nodes.
+public struct WorkflowSession {
+    public struct Identifier: Hashable {
+        private static var _nextRawID: UInt64 = 0
+        private static func _makeNextSessionID() -> UInt64 {
+            _nextRawID += 1
+            return _nextRawID
+        }
+
+        let rawIdentifier: UInt64 = Self._makeNextSessionID()
+    }
+
+    private indirect enum IndirectParent {
+        case some(WorkflowSession)
+        case none
+
+        init(_ parent: WorkflowSession?) {
+            switch parent {
+            case .some(let value):
+                self = .some(value)
+            case .none:
+                self = .none
+            }
+        }
+    }
+
+    public let workflowType: Any.Type
+
+    public let renderKey: String
+
+    public let sessionID = Identifier()
+
+    private let _indirectParent: IndirectParent
+    public var parent: WorkflowSession? {
+        switch _indirectParent {
+        case .some(let parent):
+            return parent
+        case .none:
+            return nil
+        }
+    }
+
+    init<WorkflowType: Workflow>(
+        workflow: WorkflowType,
+        renderKey: String,
+        parent: WorkflowSession?
+    ) {
+        self.workflowType = WorkflowType.self
```

Comment:
any reason for not storing the workflow itself?

Reply:
i was thinking of the session as like mostly fixed metadata about the 'node' that persists over time. the actual `Workflow` instance (that clients provide) that is associated with a given node can change over time, so if we stored that here we'd have to deal with that fact. also, this is largely modeled from the Kotlin implementation, which doesn't store the Workflow directly on this type (or, the 'props' as they're called in that context).